rfcs/0125-bootspec.md

```diff
@@ -0,0 +1,218 @@
+  The goal of this RFC is to make the feature development *easier*, not actually do it.
+- Supporting SecureBoot.
+  The authors of this RFC have done work in this regard, but this RFC is not about SecureBoot.
+- Integrating Bootspec into an existing bootloader backend is not expected to perfectly produce the exact same result. For example, it is likely that the text describing the entries in the menu may differ.
+
+# Proposed Solution
+[proposed-solution]: #proposed-solution
+
+- Each NixOS generation will have a bootspec (a JSON document) at `$out/boot.json` containing all of the boot properties for that generation.
+  NixOS's bootloader backends will read these files as inputs to the bootloader installation phase.
+  A bootspec document named `boot.json` must have a `"v1"` key that contains the body of the specification.
+  Generally, the key must follow the format of `v${schemaVersion}`.
+    - The bootloader installation phase is relatively unchanged from the way it is now.
+      The bootloader backend will have an executable that is run against a collection of generations, and the backend is any of the currently supported backends plus an "external" backend which the user can define.
+- The bootloader backends will avoid reading data from the other files and directories when possible, preferring the information in the bootspec.
+- A bootspec synthesizing tool will be used to synthesize a bootspec for generations which don't already have one.
+  This tool will be shared across all of the bootloader backends, helping produce more uniform behavior.
+- Existing bootloader backends will be updated to read properties from the bootspec, removing most if not all of their filesystem-spelunking code.
+
+### Bootspec Format v1
+[format-v1]: #format-v1
+
+
+Using the following JSON:
+
+```json5
+{
+  // Toplevel key describing the version of the specification used in the document
+  "v1": {
+
+    // Path to the stage-2 init, executed by the initrd
+    "init": "/nix/store/xxx-nixos-system-xxx/init",
+
+    // Path to the initrd
+    "initrd": "/nix/store/xxx-initrd-linux/initrd",
+
+    // Optional path to a tool to dynamically add secrets to an initrd
+    "initrdSecrets": "/nix/store/xxx-append-secrets/bin/append-initrd-secrets",
+
+    // Path to the kernel image
+    "kernel": "/nix/store/xxx-linux/bzImage",
+
+    // Kernel commandline options
+    "kernelParams": [
+      "amd_iommu=on",
+      "amd_iommu=pt",
+      "iommu=pt",
+      "kvm.ignore_msrs=1",
+      "kvm.report_ignored_msrs=0",
+      "udev.log_priority=3",
+      "systemd.unified_cgroup_hierarchy=1",
+      "loglevel=4"
+    ],
+
+    // The label of the system. It should contain the operating system, kernel version,
+    // and other user-relevant information to identify the system. This corresponds
+    // loosely to `config.system.nixos.label`.
+    "label": "NixOS 21.11.20210810.dirty (Linux 5.15.30)",
```

Comment:
Some bootloaders (GRUB for instance) also record the date when the generation was made (probably mtime of the generation symlink?). Would that be included in the label or a separate field?

Reply:
The tooling / automation is going to have to identify a collection of bootspec documents to use. It could be a directory of symlinks, or a database, or a tool run in CI -- but wherever it comes from should probably provide this sort of information.

Reply:
FWIW, I've implemented the current logic (well, an amalgamation of the current logic from different scripts) as a reusable component here: https://gitlab.com/K900/bootis/-/tree/main/nixos-list-generations